AdblockPlusSafari/FilterList+Processing.m

```diff
@@ -0,0 +1,251 @@
+                                                                           error:&error];
+  if (error != nil) {
+    return NO;
+  }
+  NSArray *matches = [regex matchesInString:expires
+                                    options:0
+                                      range:NSMakeRange(0, [expires length])];
+  NSAssert([matches count] <= 1, @"There should be at most one match");
+  NSTextCheckingResult *firstMatch = matches.firstObject;
+  if (firstMatch == nil) {
+    return NO;
+  }
+
+  *output = [[expires substringWithRange:[firstMatch range]] doubleValue] * 3600 * 24;
+  return YES;
+}
+
+// Get parser error
++ (NSError *)createParserError:(yajl_handle) hand
+{
+  NSUInteger errorBufferLength = 512;
+  uint8_t errorBuffer[errorBufferLength];
+  unsigned char *error_string = yajl_get_error(hand, 1, errorBuffer, errorBufferLength);
+  NSString *errorString = [[NSString alloc] initWithCString:(const char *)error_string encoding:NSASCIIStringEncoding];
+  return [NSError errorWithDomain:AdblockPlusErrorDomain code:0 userInfo:@{NSLocalizedDescriptionKey: errorString}];
+}
+
+- (BOOL)parseFilterListFromURL:(NSURL *__nonnull)input
+                         error:(NSError *__nullable *__nonnull)error
+{
+  NSInputStream *inputStream = [NSInputStream inputStreamWithURL:input];
+
+  yajl_handle hand = NULL;
+  AdblockPlusProcessingContext *context = [[AdblockPlusProcessingContext alloc] init];
+  void *contentPointer = (void *)CFBridgingRetain(context);
+
+  @try {
+    [inputStream open];
+
+    hand = yajl_alloc(&callbacks, NULL, contentPointer);
+    yajl_config(hand, yajl_allow_comments, 0);
+    yajl_config(hand, yajl_dont_validate_strings, 1);
+
+    // Read json file
+    const NSUInteger inputBufferLength = 256;
+    uint8_t inputBuffer[inputBufferLength];
+    NSInteger read;
+
+    while ((read = [inputStream read:inputBuffer maxLength:inputBufferLength])) {
+
+      yajl_status status = yajl_parse(hand, inputBuffer, read);
+      if (status != yajl_status_ok) {
+        *error = [[self class] createParserError:hand];
+        return NO;
+      }
+    }
+
+    // Close parser
+    yajl_status status = yajl_complete_parse(hand);
+    if (status != yajl_status_ok) {
+      *error = [[self class] createParserError:hand];
+      return NO;
+    }
+  }
+  @catch (NSException *exception) {
+    *error = [NSError errorWithDomain:AdblockPlusErrorDomain code:0 userInfo:@{NSLocalizedDescriptionKey: [exception reason]}];
+    return NO;
+  }
+  @finally {
+    CFBridgingRelease(contentPointer);
+    [inputStream close];
+    yajl_free(hand);
+  }
+
+  // JSON is atleast valid
+  if (context.filterListType == AdblockPlusFilterListTypeVersion1) {
+    self.version = nil;
+    self.expires = FilterListsUpdatePeriod;
+    return YES;
+  }
+
+  // Test if JSON is valid
+  if (!context.versionKeyFound || !context.expiresKeyFound || !context.rulesKeyFound) {
```

Comment:
I figure those headers should be optional even for version 2 filter lists. What's arguably mandatory is `rules`. `expires` should definitely not be mandatory, as described in issue 3177. `version` neither, IMO.